File: packages/core/src/lib/source-validator.ts
import { parseM3uToPlaylist } from './m3u';
import { sourceSchema, type Source } from './contracts';
import { fetchXtreamPlayerApi, isXtreamAuthSuccessful } from './xtream';

export type SourceValidationErrorCode =
  | 'invalid_url'
  | 'cors_blocked'
  | 'unreachable'
  | 'parse_error'
  | 'empty_content'
  | 'auth_failed'
  | 'unexpected_payload';

export type SourceValidationResult =
  | { ok: true; source: Source }
  | { ok: false; code: SourceValidationErrorCode; message: string };

type FetchLike = (input: string) => Promise<{ ok: boolean; status: number; text(): Promise<string> }>;

export async function validateSource(
  sourceInput: unknown,
  options: { fetcher?: FetchLike; rawM3uText?: string } = {}
): Promise<SourceValidationResult> {
  const parsed = sourceSchema.safeParse(sourceInput);
  if (!parsed.success) {
    return { ok: false, code: 'invalid_url', message: parsed.error.message };
  }
  const source = parsed.data;

  if (source.type === 'm3u_file') {
    const content = options.rawM3uText?.trim() ?? '';
    if (!content) {
      return { ok: false, code: 'empty_content', message: 'M3U file content is empty.' };
    }
    try {
      parseM3uToPlaylist(content, source.id);
      return { ok: true, source };
    } catch (error) {
      return {
        ok: false,
        code: 'parse_error',
        message: error instanceof Error ? error.message : 'Failed to parse M3U content.',
      };
    }
  }

  if (source.type === 'm3u_url') {
    const fetcher = options.fetcher ?? (async () => Promise.reject(new Error('No fetcher configured')));
    try {
      const response = await fetcher(source.url ?? '');
      if (!response.ok) {
        return {
          ok: false,
          code: 'unreachable',
          message: `Source URL returned HTTP ${response.status}.`,
        };
      }
      const text = (await response.text()).trim();
      if (!text) {
        return { ok: false, code: 'empty_content', message: 'Source URL returned empty body.' };
      }
      parseM3uToPlaylist(text, source.id);
      return { ok: true, source };
    } catch (error) {
      const message = error instanceof Error ? error.message : 'Unknown fetch error.';
      if (message.toLowerCase().includes('cors')) {
        return {
          ok: false,
          code: 'cors_blocked',
          message: 'Request was blocked by CORS. Paste raw content or import a file.',
        };
      }
      return { ok: false, code: 'unreachable', message };
    }
  }

  if (source.type === 'xtream') {
    if (!source.credentials) {
      // sourceSchema.superRefine already guards this, but narrow for TS.
      return {
        ok: false,
        code: 'invalid_url',
        message: 'Xtream sources require credentials.',
      };
    }
    const fetcher = options.fetcher ?? (async () => Promise.reject(new Error('No fetcher configured')));
    try {
      // `XtreamFetcher` only needs `.text()`; `FetchLike` is a superset.
      const xtreamFetcher = async (url: string) => {
        const r = await fetcher(url);
        return { text: () => r.text() };
      };
      const payload = await fetchXtreamPlayerApi(source.credentials, xtreamFetcher);
      if (!isXtreamAuthSuccessful(payload)) {
        return {
          ok: false,
          code: 'auth_failed',
          message: 'Xtream panel rejected the credentials (auth=0).',
        };
      }
      return { ok: true, source };
    } catch (error) {
      const message = error instanceof Error ? error.message : 'Unknown Xtream error.';
      // Zod parse failures throw `ZodError` whose message starts with `[`.
      // SyntaxError comes from JSON.parse on non-JSON bodies (e.g. HTML error
      // pages or HTTP 461 user-agent blocks that return text).
      if (
        error instanceof SyntaxError ||
        message.startsWith('[') ||
        message.toLowerCase().includes('zod')
      ) {
        return {
          ok: false,
          code: 'unexpected_payload',
          message: `Xtream panel returned an unexpected response: ${message}`,
        };
      }
      if (message.toLowerCase().includes('cors')) {
        return {
          ok: false,
          code: 'cors_blocked',
          message: 'Request was blocked by CORS.',
        };
      }
      return { ok: false, code: 'unreachable', message };
    }
  }

  return { ok: true, source };
}

File: apps/android-tv/app/src/main/java/com/iptvtavern/androidtv/domain/parser/SourceValidator.kt
package com.iptvtavern.androidtv.domain.parser

import com.iptvtavern.androidtv.domain.model.Source
import com.iptvtavern.androidtv.domain.model.SourceType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

/**
 * Source validation — Kotlin port of `packages/core/src/lib/source-validator.ts`.
 *
 * Validates a source before saving: checks URL reachability, parses the
 * M3U content, and returns a typed result.
 *
 * ## Differences from web version
 *
 * - **No CORS**: Android makes direct HTTP requests — no proxy needed,
 *   no `cors_blocked` error code.
 * - **No Xtream validation here**: Xtream auth is handled in Phase 8.
 *   This validator covers M3U URL and M3U file sources only.
 * - Uses `HttpURLConnection` (simple, no extra dependencies). Phase 3
 *   doesn't need OkHttp yet — that comes with networking in later phases.
 */

/** Error codes returned when validation fails. */
enum class SourceValidationErrorCode {
    INVALID_URL,
    UNREACHABLE,
    PARSE_ERROR,
    EMPTY_CONTENT,
}

/** Validation result — either success with the source, or failure with a code + message. */
sealed class SourceValidationResult {
    data class Success(val source: Source) : SourceValidationResult()
    data class Failure(
        val code: SourceValidationErrorCode,
        val message: String,
    ) : SourceValidationResult()
}

/**
 * Validate a source input and attempt to parse its M3U content.
 *
 * @param source      The source to validate.
 * @param rawM3uText  For `m3u_file` sources: the raw file content. Ignored for URL sources.
 * @return [SourceValidationResult.Success] with the validated source, or
 *         [SourceValidationResult.Failure] with an error code and message.
 */
suspend fun validateSource(
    source: Source,
    rawM3uText: String? = null,
): SourceValidationResult {
    return when (source.type) {
        SourceType.M3U_FILE -> validateM3uFile(source, rawM3uText)
        SourceType.M3U_URL -> validateM3uUrl(source)
        SourceType.XTREAM -> {
            // Xtream validation (auth probe) is handled in Phase 8.
            // For now, just check that credentials are present.
            if (source.credentials == null) {
                SourceValidationResult.Failure(
                    code = SourceValidationErrorCode.INVALID_URL,
                    message = "Xtream sources require credentials.",
                )
            } else {
                SourceValidationResult.Success(source)
            }
        }
    }
}

/**
 * If the source has no manually-set [Source.epgUrl], auto-populate it from
 * the M3U file's `#EXTM3U url-tvg="..."` header attribute.
 *
 * Called during validation — the raw M3U text is already in memory, so this
 * is essentially free. Manual user input always takes precedence.
 */
private fun Source.withEpgFromTvg(rawM3uText: String): Source {
    if (!epgUrl.isNullOrBlank()) return this
    return extractUrlTvg(rawM3uText)?.let { copy(epgUrl = it) } ?: this
}

private fun validateM3uFile(source: Source, rawM3uText: String?): SourceValidationResult {
    val content = rawM3uText?.trim().orEmpty()
    if (content.isEmpty()) {
        return SourceValidationResult.Failure(
            code = SourceValidationErrorCode.EMPTY_CONTENT,
            message = "M3U file content is empty.",
        )
    }
    return try {
        parseM3uToPlaylist(content, source.id)
        SourceValidationResult.Success(source.withEpgFromTvg(content))
    } catch (e: Exception) {
        SourceValidationResult.Failure(
            code = SourceValidationErrorCode.PARSE_ERROR,
            message = e.message ?: "Failed to parse M3U content.",
        )
    }
}

/**
 * Fetch and validate an M3U URL.
 *
 * Runs on [Dispatchers.IO] since it does network I/O.
 * Uses [HttpURLConnection] — the simplest JDK HTTP client, no extra deps.
 */
private suspend fun validateM3uUrl(source: Source): SourceValidationResult {
    val urlString = source.url
    if (urlString.isNullOrBlank()) {
        return SourceValidationResult.Failure(
            code = SourceValidationErrorCode.INVALID_URL,
            message = "M3U URL sources require a URL.",
        )
    }

    // Validate URL format
    try {
        URL(urlString)
    } catch (_: MalformedURLException) {
        return SourceValidationResult.Failure(
            code = SourceValidationErrorCode.INVALID_URL,
            message = "Invalid URL format.",
        )
    }

    return withContext(Dispatchers.IO) {
        try {
            val connection = URL(urlString).openConnection() as HttpURLConnection
            connection.connectTimeout = 15_000
            connection.readTimeout = 30_000
            connection.requestMethod = "GET"
            // Apply per-source User-Agent if set
            source.userAgent?.let { connection.setRequestProperty("User-Agent", it) }

            try {
                val status = connection.responseCode
                if (status !in 200..299) {
                    return@withContext SourceValidationResult.Failure(
                        code = SourceValidationErrorCode.UNREACHABLE,
                        message = "Source URL returned HTTP $status.",
                    )
                }

                val text = connection.inputStream.bufferedReader().use { it.readText() }.trim()
                if (text.isEmpty()) {
                    return@withContext SourceValidationResult.Failure(
                        code = SourceValidationErrorCode.EMPTY_CONTENT,
                        message = "Source URL returned empty body.",
                    )
                }

                // Try parsing — if it throws, content is not valid M3U
                parseM3uToPlaylist(text, source.id)
                SourceValidationResult.Success(source.withEpgFromTvg(text))
            } finally {
                connection.disconnect()
            }
        } catch (_: IOException) {
            SourceValidationResult.Failure(
                code = SourceValidationErrorCode.UNREACHABLE,
                message = "Could not reach the source URL. Check the address and your network.",
            )
        } catch (e: Exception) {
            SourceValidationResult.Failure(
                code = SourceValidationErrorCode.PARSE_ERROR,
                message = e.message ?: "Failed to parse M3U content from URL.",
            )
        }
    }
}
